### ISSUE TITLE
Signing using ECDSA with `nonce-type=1` (deterministic signing) only works if `digest` has been previously set using `EVP_PKEY_CTX_set_params`

### ISSUE THREAD

**Issue by facutuesca** [openssl/openssl#23205]
This is a behavior I ran into while trying to use ECDSA with deterministic signing. After creating a new `EVP_PKEY_CTX`, initializing it and setting the [`nonce-type`](https://www.openssl.org/docs/manmaster/man7/provider-signature.html#:~:text=%E2%80%9Cnonce%2Dtype%E2%80%9D%20(OSSL_SIGNATURE_PARAM_NONCE_TYPE)%20%3Cunsigned%20integer%3E) parameter to 1 (to use [deterministic signing](https://datatracker.ietf.org/doc/html/rfc6979)), I tried to call `EVP_PKEY_sign`:

```c
pkey = .... //Create a new EC key

sign_ctx = EVP_PKEY_CTX_new(pkey, NULL);
EVP_PKEY_sign_init(sign_ctx);

// set nonce type
OSSL_PARAM params[2];
unsigned int nonce_type = 1;
params[0] = OSSL_PARAM_construct_uint(OSSL_SIGNATURE_PARAM_NONCE_TYPE, &nonce_type);
params[1] = OSSL_PARAM_construct_end();
if (!EVP_PKEY_CTX_set_params(sign_ctx, params)) {
    goto err;
}

// .....
EVP_PKEY_sign(....);
```

This fails with the following error:
```
00106AD701000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:342:Global default library context, Algorithm ( : 0), Properties (<null>)
00106AD701000000:error:0800009E:elliptic curve routines:ecdsa_sign_setup:random number generation failed:crypto/ec/ecdsa_ossl.c:198:
```

If one sets `nonce_type=0` instead (the default value), everything works correctly.
After some debugging, I found that the failure is due to the `digest` parameter not being set. More specifically, the failure occurs in the following chain of calls:

<img width="436" alt="image" src="https://github.com/openssl/openssl/assets/5762120/e065e6b4-3727-4352-b80e-23a7dc7692ae">

In other words, in the code path for ECDSA with deterministic signing the hash algorithm must be known, which only happens if one sets it with `EVP_PKEY_CTX_set_params`:

```c
// set nonce type
OSSL_PARAM params[3];
unsigned int nonce_type = 1;
char* digest_name = "SHA256";

params[0] = OSSL_PARAM_construct_uint(OSSL_SIGNATURE_PARAM_NONCE_TYPE, &nonce_type);
params[1] = OSSL_PARAM_construct_utf8_string(OSSL_SIGNATURE_PARAM_DIGEST, digest_name, 6);
params[2] = OSSL_PARAM_construct_end();

EVP_PKEY_CTX_set_params(sign_ctx, params)
```

If there is no way around having to know the hash algorithm, then it would be nice to add a note to the documentation of `nonce-type` mentioning that one must set also `digest` when setting `nonce-type=1`.

**Comment by jamuir** [comment]
Thanks for reporting this.

In your (non-deterministic) ECDSA sign operation, SHA256 is being used as the message digest.

This is b/c when a public-key alg needs a message digest and one isn't explicitly given, then SHA256 is the default.  This is documented here:

https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_get_default_digest_nid.html

For deterministic ECDSA, as your backtrace shows, a sign operation makes use of a symmetric primitive, HMAC, and HMAC does not seem to offer a default digest function.

It would be nice to have more consistency between these two versions of ECDSA, so I am inclined to view this as a bug (rather than just a documentation issue).

@slontis : do you have an opinion on this?

**Comment by slontis** [comment]
I would tend towards documentation for now.
We are comparing the behaviour of EVP_DigestSignInit to EVP_PKEY_sign().
There is also #22672

**Comment by jamuir** [comment]
> I would tend towards documentation for now.

okay.  There are three doc files that mention `nonce-type`:

```
openssl/doc$ git grep -l 'nonce-type'
man7/EVP_SIGNATURE-DSA.pod
man7/EVP_SIGNATURE-ECDSA.pod
man7/provider-signature.pod
```

It seems like `provider-signature.pod` is the right doc file to update (even though that file says it is "primarily aimed at provider authors").

**Comment by facutuesca** [comment]
Thank you @jamuir @slontis !
